nginx: add missing nginx-mod-lua pieces from openresty

jmarcet opened the pull request:
Maintainer: Thomas Heil <heil@terminal-consulting.de> & Ansuel Smith <ansuelsmth@gmail.com>
Compile tested: master x86-64
Run tested: master x86-64

Description:
Upon the recent nginx upgrade to version 1.24.0, nginx-lua is incomplete, at least a pretty basic usage with [nginx-lua-prometheus](https://github.com/knyar/nginx-lua-prometheus) fails with the following messages:
```
$ nginx -c /etc/nginx/nginx.conf
2023/05/05 09:45:00 [alert] 12167#0: detected a LuaJIT version which is not OpenResty's; many optimizations will be disabled and performance will be compromised (see https://github.com/openresty/luajit2 for OpenResty's LuaJIT or, even better, consider using the OpenResty releases from https://openresty.org/en/download.html)
2023/05/05 09:45:00 [alert] 12167#0: failed to load the 'resty.core' module (https://github.com/openresty/lua-resty-core); ensure you are using an OpenResty release from https://openresty.org/en/download.html (reason: module 'resty.core' not found:
        no field package.preload['resty.core']
        no file './resty/core.lua'
        no file '/usr/share/luajit-2.1.0-beta3/resty/core.lua'
        no file '/usr/share/lua/resty/core.lua'
        no file '/usr/share/lua/resty/core/init.lua'
        no file '/usr/share/lua/resty/core.lua'
        no file '/usr/share/lua/resty/core/init.lua'
        no file './resty/core.so'
        no file '/usr/lib/lua/resty/core.so'
        no file '/usr/lib/lua/resty/core.so'
        no file '/usr/lib/lua/loadall.so'
        no file './resty.so'
        no file '/usr/lib/lua/resty.so'
        no file '/usr/lib/lua/resty.so'
        no file '/usr/lib/lua/loadall.so') in /etc/nginx/nginx.conf:124
```

This PR addresses the second alert, adding the missing pieces according to the [lua-nginx-module setup instructions](https://github.com/openresty/lua-nginx-module)

In addition to these changes, I've had to add `lua_package_path "/usr/lib/lua/?.lua;;";` to my `nginx.conf`, although that's also explained in the [setup instructions](https://github.com/openresty/lua-nginx-module)

I'm pushing another PR to switch luajit to openresty's fork, so we get rid of the first alert too.

Ansuel commented:
Thanks a lot for taking care of this the lua resty core are nginx module and not stuff based on lua itself?
Asking since i didn't check.

jmarcet commented:
> Thanks a lot for taking care of this the lua resty core are nginx module and not stuff based on lua itself? Asking since i didn't check.

The resty modules are all lua, they're just copied from the package source. There is an additional compiled module, the nginx developer kit, although I'm not sure what it's used for, I haven't checked. I included it to be compliant with the setup instructions.

```
❯ opkg files nginx-mod-lua
Package nginx-mod-lua (1.24.0-2) is installed on root and has the following files:
/usr/lib/lua/ngx/balancer.lua
/usr/lib/lua/ngx/base64.lua
/usr/lib/lua/ngx/errlog.lua
/usr/lib/lua/ngx/ocsp.lua
/usr/lib/lua/ngx/pipe.lua
/usr/lib/lua/ngx/process.lua
/usr/lib/lua/ngx/re.lua
/usr/lib/lua/ngx/req.lua
/usr/lib/lua/ngx/resp.lua
/usr/lib/lua/ngx/semaphore.lua
/usr/lib/lua/ngx/ssl.lua
/usr/lib/lua/ngx/ssl/clienthello.lua
/usr/lib/lua/ngx/ssl/session.lua
/usr/lib/lua/resty/core.lua
/usr/lib/lua/resty/core/base.lua
/usr/lib/lua/resty/core/base64.lua
/usr/lib/lua/resty/core/coroutine.lua
/usr/lib/lua/resty/core/ctx.lua
/usr/lib/lua/resty/core/exit.lua
/usr/lib/lua/resty/core/hash.lua
/usr/lib/lua/resty/core/misc.lua
/usr/lib/lua/resty/core/ndk.lua
/usr/lib/lua/resty/core/param.lua
/usr/lib/lua/resty/core/phase.lua
/usr/lib/lua/resty/core/regex.lua
/usr/lib/lua/resty/core/request.lua
/usr/lib/lua/resty/core/response.lua
/usr/lib/lua/resty/core/shdict.lua
/usr/lib/lua/resty/core/socket.lua
/usr/lib/lua/resty/core/time.lua
/usr/lib/lua/resty/core/uri.lua
/usr/lib/lua/resty/core/utils.lua
/usr/lib/lua/resty/core/var.lua
/usr/lib/lua/resty/core/worker.lua
/usr/lib/lua/resty/lrucache.lua
/usr/lib/lua/resty/lrucache/pureffi.lua
/usr/lib/nginx/modules/ndk_http_module.so
/usr/lib/nginx/modules/ngx_http_lua_module.so
```

Ansuel commented:
If they are la module then we may consider putting them in a separate
makefile instead of adding them in nginx?

Il Lun 15 Mag 2023, 10:22 Javier Marcet ***@***.***> ha
scritto:

> Thanks a lot for taking care of this the lua resty core are nginx module
> and not stuff based on lua itself? Asking since i didn't check.
>
> The resty modules are all lua, they're just copied from the package
> source. There is an additional compiled module, the nginx developer kit,
> although I'm not sure what it's used for, I haven't checked. I included it
> to be compliant with the setup instructions.
>
> ❯ opkg files nginx-mod-lua
> Package nginx-mod-lua (1.24.0-2) is installed on root and has the following files:
> /usr/lib/lua/ngx/balancer.lua
> /usr/lib/lua/ngx/base64.lua
> /usr/lib/lua/ngx/errlog.lua
> /usr/lib/lua/ngx/ocsp.lua
> /usr/lib/lua/ngx/pipe.lua
> /usr/lib/lua/ngx/process.lua
> /usr/lib/lua/ngx/re.lua
> /usr/lib/lua/ngx/req.lua
> /usr/lib/lua/ngx/resp.lua
> /usr/lib/lua/ngx/semaphore.lua
> /usr/lib/lua/ngx/ssl.lua
> /usr/lib/lua/ngx/ssl/clienthello.lua
> /usr/lib/lua/ngx/ssl/session.lua
> /usr/lib/lua/resty/core.lua
> /usr/lib/lua/resty/core/base.lua
> /usr/lib/lua/resty/core/base64.lua
> /usr/lib/lua/resty/core/coroutine.lua
> /usr/lib/lua/resty/core/ctx.lua
> /usr/lib/lua/resty/core/exit.lua
> /usr/lib/lua/resty/core/hash.lua
> /usr/lib/lua/resty/core/misc.lua
> /usr/lib/lua/resty/core/ndk.lua
> /usr/lib/lua/resty/core/param.lua
> /usr/lib/lua/resty/core/phase.lua
> /usr/lib/lua/resty/core/regex.lua
> /usr/lib/lua/resty/core/request.lua
> /usr/lib/lua/resty/core/response.lua
> /usr/lib/lua/resty/core/shdict.lua
> /usr/lib/lua/resty/core/socket.lua
> /usr/lib/lua/resty/core/time.lua
> /usr/lib/lua/resty/core/uri.lua
> /usr/lib/lua/resty/core/utils.lua
> /usr/lib/lua/resty/core/var.lua
> /usr/lib/lua/resty/core/worker.lua
> /usr/lib/lua/resty/lrucache.lua
> /usr/lib/lua/resty/lrucache/pureffi.lua
> /usr/lib/nginx/modules/ndk_http_module.so
> /usr/lib/nginx/modules/ngx_http_lua_module.so
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/packages/pull/21018#issuecomment-1547408025>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQVHY7YESQZPH2DR2M3XGHRV3ANCNFSM6AAAAAAYBJ5KEI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

jmarcet commented:
Well, I thought about that, but this belongs to `nginx-mod-lua` not `nginx` itself. Not only that, in practical terms the former is not complete.

How would you split them? I mean, shall they be two separate packages? One? Three if we also count the `ndk`. Plus, how should they be named?

Moreover, shall they be dependencies of `nginx-mod-lua`? Or should they be offered as new configuration options?

Overall I think the cleanest approach is just adding the missing pieces to `nginx-mod-lua`, although that is IMHO.

Ansuel commented:
ndk can't be split... but if lua module for openresty can live outside nginx package then I would just create 2 new package and add them as a dependency (again if this is possible) the intention is to not bloat the nginx package too much

jmarcet commented:
I've left ndk within `nginx-mod-lua` and split the resty modules into two additional `nginx-mod-lua-resty-core` & `nginx-mod-lua-resty-lrucache` modules

jmarcet commented:
There's some issue in the dependencies I cannot find how to solve:
```
❯ make oldconfig
Collecting package info: done
tmp/.config-package.in:112508:error: recursive dependency detected!
tmp/.config-package.in:112508:  symbol PACKAGE_nginx-mod-lua-resty-lrucache depends on PACKAGE_nginx-mod-lua-resty-lrucache
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:112408:error: recursive dependency detected!
tmp/.config-package.in:112408:  symbol PACKAGE_nginx-full depends on PACKAGE_nginx-full
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:112484:error: recursive dependency detected!
tmp/.config-package.in:112484:  symbol PACKAGE_nginx-mod-lua depends on PACKAGE_nginx-mod-lua
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:112496:error: recursive dependency detected!
tmp/.config-package.in:112496:  symbol PACKAGE_nginx-mod-lua-resty-core depends on PACKAGE_nginx-mod-lua-resty-core
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

#
# No change to .config
#
```

Ansuel commented:
@jmarcet Hi nginx makefile was refactored, can you update this with new implementation? Should be easier to add modules.

jmarcet commented:
@Ansuel

Sorry for my absence lately.

I've refactored it, although it might not be the cleanest way. I also dropped the `ndk` module, since it is not an `nginx-mod` I wasn't sure how to add it cleanly and I have no use for it, so I am not testing it at all.

jmarcet commented:
@Ansuel

Ansuel commented on net/nginx/Makefile:
```diff
@@ -207,14 +209,40 @@ define Download/nginx-mod-geoip2
   PROTO:=git
 endef
 
-
```
please drop this change

Ansuel commented on net/nginx/Makefile:
```diff
@@ -207,14 +209,40 @@ define Download/nginx-mod-geoip2
+
+define Package/nginx-mod-lua-resty-lrucache
+  $(call Package/nginx/default)
+  DEPENDS:=+nginx-mod-lua
```
ass luajit2 as a dependency

Ansuel commented on net/nginx/Makefile:
```diff
@@ -207,14 +209,40 @@ define Download/nginx-mod-geoip2
 
+define Package/nginx-mod-lua-resty-core
+  $(call Package/nginx/default)
+  DEPENDS:=+nginx-mod-lua
```
add lua-resty-lrucache as a dependency

jmarcet replied:
Adding also `+luajit2` gets rid of:

```
tmp/.config-package.in:24994:error: recursive dependency detected!
tmp/.config-package.in:24994:   symbol PACKAGE_luajit2 is selected by PACKAGE_nginx-mod-lua-resty-lrucache
tmp/.config-package.in:114130:  symbol PACKAGE_nginx-mod-lua-resty-lrucache is selected by PACKAGE_nginx-mod-lua-resty-core
tmp/.config-package.in:114116:  symbol PACKAGE_nginx-mod-lua-resty-core depends on PACKAGE_luajit2
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"
```

Ansuel replied:
actually in theory -mod-lua should change the dependency to luajit2

Ansuel commented:
the recursive dependency error needs to be solved and we need to wait for luajit2 to be merged for this.

jmarcet commented:
> the recursive dependency error needs to be solved and we need to wait for luajit2 to be merged for this.

I still get these:
```
tmp/.config-package.in:114103:error: recursive dependency detected!
tmp/.config-package.in:114103:  symbol PACKAGE_nginx-mod-lua depends on PACKAGE_nginx-mod-lua
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

tmp/.config-package.in:25028:error: recursive dependency detected!
tmp/.config-package.in:25028:   symbol PACKAGE_lualogging depends on PACKAGE_lualogging
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"
```

BKPepe commented:
I will fix during the weekend the Luajit2 for PowerPC. I know what's going on. :) It does not include the relevant changes, which was merged to its repo recently.

Ansuel commented:
@BKPepe thanks! Aside from that other test are failing due to old naxsi module and I just fixed it.
